### PR TITLE
Update in Multimodel plot functionality

### PR DESCRIPTION
**Pull request by curadomr**
- getPlottingData provides mapping object, only for provided feature_id.
- testID and modelID must be vectors of same lenght for multimodel plots.
- results for multiple tests from a given model are provided as nested lists in plottingData, within the list element 'result', and named after the corresponding testID
- testPlot.R and vignette/OmicNavigatorAPI.Rnw adapted accordingly.

### CODE REVIEW

**Comment by jdblischak** [comment]
Quick update from our discussion: the plan for the mapping object is to be a list of data frames (one per modelID), similar to features and samples. In most cases there will only be a single mapping table to map featureIDs across all modelIDs of the study. Thus a user can choose the modelID "default" when adding it with `addMapping()`. In the rare case where a user wants different mappings based on the currently selected modelID, they could specify a different mapping table per modelID.

File: R/plots.R [hunk 1]
```diff
@@ -268,25 +311,17 @@ getPlottingData <- function(study, modelID, featureID, testID = NULL, libraries
...
-    mappingdf <- as.data.frame(mapping, stringsAsFactors = FALSE)
-    mapping_features_all <- stats::na.omit(mappingdf)
-    column_order <- unique(c(modelID[1], colnames(mapping_features_all)))
-    mapping_features_all <- mapping_features_all[, column_order]
-
-    mapping_features <- mapping_features_all[which(mapping_features_all[,1] %in% featureID),]
-
-    testID_all <- testID
+    mappingPlottingData <- getMappingPlottingData(study, modelID, featureID, testID, libraries)
+    list2env(mappingPlottingData, environment())
```
Comment (jdblischak) [inline R/plots.R:315]: Please don't use `list2env()`. This makes it very hard to reason about. `R CMD check` threw the following NOTE:

```
* checking R code for possible problems ... NOTE
getPlottingData: no visible binding for global variable
  'mapping_features'
getPlottingData: no visible binding for global variable 'testID_all'
Undefined global functions or variables:
  mapping_features testID_all
```

I think `mapping_features` and `testID_all` are created by the call to `list2env()`. To make it easier to follow the code, please explicitly assign these variables:

```R
mapping_features <- mappingPlottingData$mapping_features
testID_all <- mappingPlottingData$testID_all
modelID <- mappingPlottingData$modelID
```

This also alerts the reader to the fact that `modelID` has been modified (looks like it is sorted in `getMappingPlottingData`)

File: vignettes/OmicNavigatorAPI.Rnw [hunk 2]
```diff
@@ -275,13 +275,13 @@ plotStudy(
 Finally, the study can provide ``multiModel'' plots, which accept more than
 one modelID and one or more testIDs per modelID. These can be provided via the
 field \texttt{plotType} in the output from \texttt{listStudies()} (Section
-\ref{sec:list-studies}). Note that for plotType = ``multiModel'', testID must be
-provided as a named vector, with each testID named after the related modelID
+\ref{sec:list-studies}). Note that for plotType = ``multiModel'', testID and
+modelID should be vectors of same length, where the index position indicate
+which test in testID relate to which model in modelID.
 
 <<plotStudy-plotMultiTest, fig=TRUE>>=
```
Comment (jdblischak) [inline vignettes/OmicNavigatorAPI.Rnw:282]: Please update the label to `plotStudy-plotMultiModel`

**Review by jdblischak — CHANGES_REQUESTED** [review]
Great work! It's coming along. Please also bump the version to 1.12.0 and add a note to `NEWS.md`

**Comment by jdblischak** [comment]
@curadomr Before you make additional changes, make sure to pull my commits to your computer first:

```
git pull origin multimodel_mapping
```

File: R/check.R [hunk 3]
```diff
@@ -413,30 +413,25 @@ checkPlots <- function(plots) {
...
-      featAligned <- any(apply(!sapply(tempModel, is.na), 1, sum) > 1)
-    } else {
-      featAligned <- any(sum(!sapply(tempModel, is.na)) > 1)
-    }
-
-    if (!is.na(featAligned) && !featAligned) {
-      stop(sprintf("Model \"%s\" does not present any feature mapping to another model.", colnames(mappingdf)[[i]]))
+  for (i in seq_along(mapping)) {
+    # stop if mapping object contains anything else besides data frames
+    stopifnot("mapping object must be a list of at least one data frame" = inherits(mapping[[i]], "data.frame"))
```
Comment (jdblischak) [inline R/check.R:418]: This new feature to customize the error messages was added in [R 4.0.0](https://cran.r-project.org/doc/manuals/r-release/NEWS.html):

> `stopifnot()` now allows customizing error messages via argument names, thanks
> to a patch proposal by Neal Fultz in PR#17688.

To support previous versions of R, please change these 3 instances of `stopifnot()` to instead to use the pattern:

```R
if (!) {
  stop()
}
```

eg

```R
if (!inherits(mapping[[i]], "data.frame")) {
  stop("mapping object must be a list of at least one data frame")
}
```

File: R/plots.R (hunk 1 above)
Comment (curadomr) [reply R/plots.R]: Updated. List2env is replaced by explicit variable attribution, as indicated.

File: R/check.R (hunk 3 above)
Comment (curadomr) [reply R/check.R]: Updated accordingly.

**Comment by jdblischak** [comment]
I figured out how to fix the errors related to missing values in character columns. I'm going to merge this first and then push my fix.

**Review by jdblischak — APPROVED** [review]

**Comment by jdblischak** [comment]
Thanks @curadomr!

**Comment by jdblischak** [comment]
Support for missing values in character columns added in bc3a831992168c71d0df531462ec51647f2dbda9

**Pull request merged**